### ISSUE TITLE
UI

### ISSUE THREAD

**Issue by ArxdSilva** [ArxdSilva/XML-Comp#10]

**Comment by dixonwille** [comment]
@ArxdSilva Golang doesn't really have a good Cross platform UI. Either create a web server that is ran or cli menu to accomplish your needs. Want to know your thoughts.

**Comment by ArxdSilva** [comment]
@dixonwille I was thinking about something like this image bellow from [go-gtk](http://mattn.github.io/go-gtk/), but a local webserver isn't a bad idea, the problem only is not to let the user run it in background forever after It's done (easily fixable by adding a termination button).

![screenshot](https://cloud.githubusercontent.com/assets/9372503/19736845/d3bfa82e-9b8f-11e6-86d8-2bd87b90248b.png)

**Comment by dixonwille** [comment]
@ArxdSilva maybe use [electron](http://electron.atom.io/)? I might be able to find a way to run and stop the "Web Server" while it is running?

**Comment by ArxdSilva** [comment]
@dixonwille that would be perfect

**Comment by dixonwille** [comment]
So templated out the electron side just need more information on what you want for API and if you want to always write to a file?

**Comment by dixonwille** [comment]
Food For Though:

* Textbox for Original Folder Location
* Textbox for Translate Folder Location
* Both would have a browse button for native file explorer
* Have a run button

After run:

* Two column view (similar to git diff if we can)
* Show what tag exists in one not the other
  * Need to think of the case if a tag was removed.
* This way it does not matter which file comes first.

If we agree on these turns I can make the GUI better In One PR. Have a second PR for creating API server. And a third to tie them together!

**Comment by dixonwille** [comment]
@ArxdSilva just saw your original idea and I think the submit button should show the changes and not save to a file. That way user knows it is done. Also doesn't make the user leave the application to view changes.

**Comment by ArxdSilva** [comment]
@dixonwille I see where you are heading into, you're thinking about a more focused folder work comparison, so It will display the diff. My only concern with this is when It get's too big (isnt a big problem at all). 

I was thinking on something more authomated to run once and get all files & folders diffs, but I believe that your idea is much more user friendly. We can do both, mine focused on CLI and yours on single folder-files diff.

Anyway the user will jump into his/her text editor to change/create the missing folders-files-tags.

**Comment by dixonwille** [comment]
So I may have a way to view lots of changes. Keep things separated, Drawing up designs now.

**Comment by ArxdSilva** [comment]
The only problem is that the `tags diff` func is a WIP x.x

**Comment by dixonwille** [comment]
Wonderful things about API's you only connect what you need! Seperate your endpoints to folders, files, tags. Then just call all three to get your information!

**Comment by dixonwille** [comment]
Here is what I came up with. I can make it look prettier when programming (not at all a designer). Thoughts or questions?

NOTES:

* Browse and Download button open dialog to select folder
* There should be a back button on the second screen
* On second screen the black text means no change
* The green text means that this has it but other does not
* Red text means it doesn't have what the other does
* The arrows are to allow scrolling. Expanding an arrow on one side will also expand it on the other side.

![main](https://cloud.githubusercontent.com/assets/6287275/19777700/768546ec-9c47-11e6-8699-601f20b445fa.png)
![submit](https://cloud.githubusercontent.com/assets/6287275/19777694/72278c72-9c47-11e6-8b3f-d5e3afcd9068.png)

**Comment by ArxdSilva** [comment]
@dixonwille now with v0.0.4 I think we can move on with this, I'll keep this updated as I make some changes, your help is apreciated (:

**Comment by dixonwille** [comment]
@ArxdSilva just let me know what you want, when you want. And I will make it happen?

**Comment by ArxdSilva** [comment]
@dixonwille I agree on what you have designed, the only difference is that as I wanted to make a Minimum Viable Product ASAP I'm just changing how versions will be, so for now I want to make this viable only on same folder, without the sub sequential diffs. I've already done the `tags diff`  that I've mentioned before, so we should not have any problem to integrate It into the electron API.

**Comment by dixonwille** [comment]
👍 That is something my company is trying to teach me. Haha, I love to blow things out of the water when all they need is just something working and adding features later! I'll get the first screen working (no submit just download button) then we can add the submit button in later! Sound good or am I missunderstanding what you want for the Minimum?

**Comment by ArxdSilva** [comment]
@dixonwille alright, I'm anxious to see what you will come up with 👍

**Comment by dixonwille** [comment]
@ArxdSilva Slowly working on the GUI. Don't have a server yet to communicate with but I am trying to get as much of the GUI functioning then work on plugging it up. [Compare](https://github.com/ArxdSilva/XML-Comp/compare/master...dixonwille:master)

TODO:
* Change the menu
* Find a way to distribute this package with ease
* Test on Mac and Linux (developing in Windows currently and I do have access to the others)
* Make a Guidlines file to go along with the MIT license

To run simply pull my changes, run `npm install`, and then `npm run`. This will start the GUI. Browse buttons work (allows a user to choose a folder). Download button disabled if checkbox not checked. The license and guidline link opens a new window to show the license and guidlines. Let me know if something needs to change.

**Comment by ArxdSilva** [comment]
@dixonwille , thanks for all your effort! I only develop on mac (cause I'm a beast using windows) and got this msg after running `npm run`:
```shell
Lifecycle scripts included in xml-comp:
  start
    electron .
  install
    bower install

available via `npm run-script`:
  dev
    cross-env DEVELOP=true electron .
```
It seems to be related with your package.json scripts,  all the dependencies were installed properly. BTW I'll try on Ubuntu later.

**Comment by dixonwille** [comment]
wrong command I told you sorry.... `npm run dev`. The above was asking what to run. If you would prefer to see what most people see (without dev tools) run `npm start`. Sorry about that.

**Comment by ArxdSilva** [comment]
@dixonwille that's nice! Just did my setup on windows and It's working perfectly as you said.

BTW I'll try to list this repo on [awesome-go](https://github.com/avelino/awesome-go) so I may need to move the client to a new repo and vendor this one (only if they only accept 100% go code); I'll get in touch with their devs and find It out, but dont worry.

**Comment by dixonwille** [comment]
I have 2 packages on there and they do critique before posting it there. But they are helpful critiques and should not be taken as bad mistakes!!! They are very friendly and such. Just let me know what you plan to do. It will make the project cleaner if you do move the GUI out. May be worth creating an organization called XML-Comp then have your go package named core, and this GUI as GUI or something. Just a suggestion?
PS Organizations are free as long as everything is open source so that is cool!

**Comment by dixonwille** [comment]
@ArxdSilva do you have an idea for a logo? I can try and sketch one up for you?

**Comment by ArxdSilva** [comment]
@dixonwille done, now we can think separately on CLI/GUI. About the logo, just need to be clean and simple, I added some random that I was able to do in 2 min, but It would be nice to have a decent logo.

**Comment by dixonwille** [comment]
<img width="512" alt="xmlcompare" src="https://cloud.githubusercontent.com/assets/6287275/21082656/ff3ef60c-bfad-11e6-969b-1a39fb52d0cf.png">

Not a designer but this is what I came up with @ArxdSilva

**Comment by ArxdSilva** [comment]
@dixonwille got this from a website
![comp](https://cloud.githubusercontent.com/assets/9372503/21082806/18a167e4-bfca-11e6-816a-e9930637c845.png)

**Comment by dixonwille** [comment]
@ArxdSilva do you like the way it looks at 16x16 px because that is about the smallest an icon will get? For a logo though that looks great not so much for an Icon? may make the brackets closer together?

**Comment by ArxdSilva** [comment]
@dixonwille yeah it sucks as icon

**Comment by dixonwille** [comment]
I may be able to replicate it with Affinity Designer then edit the main details into an icon?

**Comment by ArxdSilva** [comment]
@dixonwille dont worry, LGTM the way It is now, It's better now to focus on coding.

**Comment by ArxdSilva** [comment]
this should help us out a bit: https://electron.atom.io/docs/api/shell/

**Comment by dixonwille** [comment]
@ArxdSilva what it seems like is your second checkbox should be able to be covered by the link you described. As for your first you have two options

1. Make a server executable that accepts HTTP and JSON
2. Just call the CLI like you would manually

Either way you would probably need to use something similar to (https://nodejs.org/api/child_process.html#child_process_child_process)

**Comment by ArxdSilva** [comment]
@dixonwille do you think that It'd be nice to have a binary versioned on gui?

**Comment by dixonwille** [comment]
@ArxdSilva I do think it would be nice.
